docs/src/index.md

```diff
@@ -4,6 +4,349 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+    At the end of the day, they are all _wiggles_ or _wobbles_.
+
+### Self derivative `Δself`, `∂self` etc.
+
+!!! terminology
+    To my knowledge there is no standard terminology for this.
+    Other good names might be `Δinternal`/`∂internal`.
+
+From the mathematical perspective, one may have been wondering what all this `Δself`, `∂self` is.
+After all a function with two inputs, say `f(a, b)`, only has two partial derivatives:
+``\dfrac{∂f}{∂a}``, ``\dfrac{∂f}{∂b}``.
+Why then does a `pushforward` take in this extra `Δself`, and why does a `pullback` return this extra `∂self`?
+
+The reason is that in Julia the function `f` may itself have internal fields.
+For example a closure has the fields it closes over; a callable object (i.e. a functor) like a `Flux.Dense` has the fields of that object.
+
+**Thus every function is treated as having the extra implicit argument `self`, which captures those fields.**
+So every `pushforward` takes in an extra argument, which is ignored unless the original function had fields.
+In is common to write `function foo_pushforward(_, Δargs...)` in the case when `foo` does not have fields.
+Similarly every `pullback` return an extra `∂self`, which for things without fields is the constant `NO_FIELDS`, indicating there are no fields within the function itself.
+
+#### Pushforward / Pullback summary
+
+- **Pushforward:**
+    - returned by `ffrule`
+    - takes input space wiggles, gives output space wobbles
+    - 1 argument per original function argument + 1 for the function itself
+    - 1 return per original function return
+
+- **Pullback**
+   - returned by `rrule`
+   - takes output space wobbles, gives input space wiggles
+   - 1 argument per original function return
+   - 1 return per original function argument + 1 for the function itself
+
+#### Pushforward/Pullback and Total Derivative/Gradient
+
+The most trivial use of `frule` and returned `pushforward` is to calculate the [Total Derivative](https://en.wikipedia.org/wiki/Total_derivative):
+
+```julia
+y, f_pushforward = frule(f, a, b, c)
+ẏ = f_pushforward(1, 1, 1, 1)  # for appropriate `1`-like perturbation.
+```
+
+Then we have that `ẏ` is the _total derivative_ of `f` at `(a, b, c)`, written mathematically as ``df_{(a,b,c)}``
+
+Similarly, the most trivial use of `rrule` and returned `pullback` is to calculate the [Gradient](https://en.wikipedia.org/wiki/Gradient):
+
+```julia
+y, f_pullback = rrule(f, a, b, c)
+∇f = f_pullback(1)  # for appropriate `1`-like seed.
+s̄, ā, b̄, c̄ = ∇f
+```
+Then we have that `∇f` is the _gradient_ of `f` at `(a, b, c)`.
+And we thus have the partial derivatives ``f̄ = \dfrac{∂f}{∂f}``, ``ā` = \dfrac{∂f}{∂a}``, ``b̄ = \dfrac{∂f}{∂b}``, ``c̄ = \dfrac{∂f}{∂c}``, including the and the self-partial derivative, ``f̄``.
```

Comment:
```suggestion
And we thus have the partial derivatives ``f̄ = \dfrac{∂f}{∂f}``, ``ā` = \dfrac{∂f}{∂a}``, ``b̄ = \dfrac{∂f}{∂b}``, ``c̄ = \dfrac{∂f}{∂c}``, including the self-partial derivative, ``f̄``.
```

Reply:
`f̄` stated twice

Reply:
``f̄ = \dfrac{∂f}{∂f}`` looks like it should be always 1, I think you should use a different symbol for internal fields of the function (maybe s)?

Reply:
Yeah, I was thinking `\phi`
but for now I say self.
Mostly.
So this is wrong for sure.

Reply:
```suggestion
And we thus have the partial derivatives ``s̄elf, = \dfrac{∂f}{∂s̄elf}``, ``ā` = \dfrac{∂f}{∂a}``, ``b̄ = \dfrac{∂f}{∂b}``, ``c̄ = \dfrac{∂f}{∂c}``, including the and the self-partial derivative, ``s̄elf,``.
```